drivers/include/adc_ng.h

```diff
@@ -0,0 +1,242 @@
+ *
+ * @pre     The ADC @p adc is initialized using @ref adc_ng_init
+ *
+ * @retval  0               Success
+ * @retval  <0              A device specific error occurred
+ */
+static inline int adc_ng_single(uint8_t adc, uint32_t *dest)
+{
+    assert(adc < ADC_NG_NUMOF);
+    assert(dest);
+    return adc_ng_drivers[adc]->single(adc_ng_handles[adc], dest);
+}
+
+/**
+ * @brief   Perform a burst conversion using the specified ADC
+ *
+ * @param[in]       adc     ADC device to use
+ * @param[out]      dest    The results are stored here
+ * @param[in]       num     The number of conversions to perform
+ *
+ * @pre     The ADC @p adc is initialized using @ref adc_ng_init
+ *
+ * @retval  0               Success
+ * @retval  <0              A device specific error occurred
+ *
+ * With `USEMODULE += adc_ng_burst`, some ADC drivers might provide a highly
+ * efficient implementation e.g. using DMA. If either the driver does not
+ * provide such implementation, or the module `adc_ng_burst` is not used, a
+ * slower but ROM-efficient fallback implementation is used instead.
+ */
+int adc_ng_burst(uint8_t adc, uint32_t *dest, size_t num);
+
+/**
+ * @brief   Initialize an ADC channel, perform a single conversion with
+ *          maximum resolution and range, and power it off again
+ *
+ * @param[in]       adc     ADC device to use
+ * @param[in]       chan    Channel to use
+ * @param[out]      dest    The result is stored here
+ *
+ * @retval  0               Success
+ * @retval  <0              A driver specific error occurred
+ *
+ * Refer to the documentation of @ref adc_init for details on @p ref
+ */
+static inline int adc_ng_quick(uint8_t adc, uint8_t chan,
+                               uint32_t *dest)
+{
+    uint16_t ref = ADC_NG_MAX_REF;
+    int retval = adc_ng_init(adc, chan, adc_ng_max_res(adc), &ref);
+    if (retval) {
+        return retval;
+    }
+    retval = adc_ng_single(adc, dest);
+    adc_ng_off(adc);
+    return retval;
+}
+
+/** @} */
+
+/**
+ * @name    Functions to get measurements in voltage levels
+ *
+ * @{
+ */
+
+/**
+ * @brief   Run a single measurements and get the result in mV
+ *
+ * @param[in]       adc     ADC to use
+ * @param[out]      dest_mv Write the measurement result in mV here
+ *
+ * @pre     The state given in @p state has been initialized (see
+ *          @ref adc_ng_init) and the ADC currently is powered
+ *
+ * @retval  0               Success
+ * @retval  <0              A driver specific error occurred
+ *
+ * @note    Please note that a while a precision roughly in the order of µV
+ *          is with the right setup and ADC technically feasible, the accuracy
+ *          of the result is more likely in the order of mV.
+ */
+static inline int adc_ng_voltage(uint8_t adc, uint16_t *dest_mv)
+{
+    uint32_t sample;
+    int retval = adc_ng_single(adc, &sample);
+    if (retval) {
+        return retval;
+    }
+
+    *dest_mv = adc_ng_convert(adc, sample);
+    return 0;
+}
+/** @} */
+
```

Comment:
It might be useful to have a function which returns the number of the channels an ADC has.
```suggestion
/**
 * @brief   Return the number of channels of an ADC
 *
 * @param[in]       adc     ADC to use
 *
 * @retval  the number of channels
 * @retval  -ENOTSUP  not supported
static inline int adc_ng_ch_numof(const ads101x_t *dev)
{
    assert(adc < ADC_NG_NUMOF);
    assert(dest);
    if (adc_ng_drivers[adc]->ch_numof) {
        return adc_ng_drivers[adc]->ch_numof();
    }
    else {
        return -ENOTSUP
    }
}
```

Reply:
Agreed, that is certainly useful.

But that is needs some precise agreement on how the number is to be interpreted. E.g. the special channels for the internal NTC (if supported) and for entropy might be counted as well, or not.

Some ADCs might have not all internal inputs broken out as pins. E.g. lets say a cheaper MCU only has ADC inputs 0, 1, 2, 5, 6, 7. Should the result be the number of the highest available channel (7), should it be the number of channels including holes (8), or should it be the number of actually supported channels (6)?

Right now I also would like to allow the driver to expose more "special" channels, e.g. like the differential inputs on the bigger ATmega ADCs, by mapping them to high channel numbers. But those are very device specific, so I would like to take here only the normal "single ended inputs" into account for the number. People wanting to use special features of on ADC that is not consistently available on most other ADCs will need to consult the documentation of that specific ADC backend in any case. And people only wanting to use the standard features might get confused if the special channels show up here as well.

Reply:
> Some ADCs might have not all internal inputs broken out as pins. E.g. lets say a cheaper MCU only has ADC inputs 0, 1, 2, 5, 6, 7. Should the result be the number of the highest available channel (7), should it be the number of channels including holes (8), or should it be the number of actually supported channels (6)?

Hm, good question. The most correct answer would probably be the number of channels actually supported. But as with the highest available channel, someone might try to iterate over the number of channels, which would lead to errors in any case. For example, `tests/periph_adc` simply iterates over all available ADC lines.

Maybe it is not a good idea to have such a function after all. Otherwise, incorrect use is possible.

Reply:
When, the init function will return `-ENXIO` for unsupported channels.

I would like to aim for this:
- The channel numbers in the range [0; *n*] must refer to single ended channels, where *n* is the number of the ADC input with the highest number
- Channel 224 and above (check with `(channum & 0xe0) == 0xe0`)  are reserved for special channels, such as the entropy channel or the internally connected NTC channel
- Channels in the range ]*n*; 224[ are backend specific channels, e.g. differential channels with two inputs.
- The backend is free to return `-ENXIO` for every channel number, except *n*. (E.g. if the data sheet names the first input not 0, but 1, then requesting channel 0 should return `-ENXIO`. Likewise for holes.)

With that defined, a function returning the number of the highest supported single ended channel (*n*) would be quite convenient for special use cases. When iterating over all single ended channels the user would start with 0 and stop with *n* - skipping all `-ENXIO` channels in between.

Reply:
OK, this should be implemented and documented accordingly. The function name got rather lengthy... I'm really bad at finding short but precise names.